crates/analyzer/src/namespace/items.rs

```diff
@@ -1324,6 +1489,292 @@ impl StructFieldId {
     }
 }
 
+#[derive(Debug, PartialEq, Eq, Hash, Clone)]
+pub struct Impl {
+    pub trait_id: TraitId,
+    pub receiver: Type,
+    pub module: ModuleId,
+    pub ast: Node<ast::Impl>,
+}
+
+#[derive(Default, Debug, PartialEq, Eq, PartialOrd, Ord, Hash, Copy, Clone)]
+pub struct ImplId(pub(crate) u32);
+impl_intern_key!(ImplId);
+impl ImplId {
+    pub fn data(&self, db: &dyn AnalyzerDb) -> Rc<Impl> {
+        db.lookup_intern_impl(*self)
+    }
+    pub fn span(&self, db: &dyn AnalyzerDb) -> Span {
+        self.data(db).ast.span
+    }
+
+    pub fn module(&self, db: &dyn AnalyzerDb) -> ModuleId {
+        self.data(db).module
+    }
+
+    pub fn all_functions(&self, db: &dyn AnalyzerDb) -> Rc<[FunctionId]> {
+        db.impl_all_functions(*self)
+    }
+
+    pub fn trait_id(&self, db: &dyn AnalyzerDb) -> TraitId {
+        self.data(db).trait_id
+    }
+
+    pub fn receiver(&self, db: &dyn AnalyzerDb) -> Type {
+        self.data(db).receiver.clone()
+    }
+
+    pub fn ast(&self, db: &dyn AnalyzerDb) -> Node<ast::Impl> {
+        self.data(db).ast.clone()
+    }
+
+    pub fn functions(&self, db: &dyn AnalyzerDb) -> Rc<IndexMap<SmolStr, FunctionId>> {
+        db.impl_function_map(*self).value
+    }
+    pub fn function(&self, db: &dyn AnalyzerDb, name: &str) -> Option<FunctionId> {
+        self.functions(db).get(name).copied()
+    }
+    pub fn parent(&self, db: &dyn AnalyzerDb) -> Item {
+        Item::Module(self.data(db).module)
+    }
+
+    fn validate_type_or_trait_is_in_ingot(
+        &self,
+        db: &dyn AnalyzerDb,
+        sink: &mut impl DiagnosticSink,
+        type_module: Option<ModuleId>,
+    ) {
+        let impl_module = self.data(db).module;
+        let is_allowed = match type_module {
+            None => impl_module == self.data(db).trait_id.module(db),
+            Some(val) => val == impl_module || self.data(db).trait_id.module(db) == impl_module,
+        };
+
+        if !is_allowed {
+            sink.push(&errors::fancy_error(
+                "illegal `impl`. Either type or trait must be in the same ingot as the `impl`",
+                vec![Label::primary(
+                    self.data(db).ast.span,
+                    format!(
+                        "Neither `{}` nor `{}` are in the ingot of the `impl` block",
+                        self.data(db).receiver,
+                        self.data(db).trait_id.name(db)
+                    ),
+                )],
+                vec![],
+            ))
+        }
+    }
+
+    pub fn sink_diagnostics(&self, db: &dyn AnalyzerDb, sink: &mut impl DiagnosticSink) {
```

Comment:
It seems trait visibility is not considered, should we validate it in this function?

Reply:
Eagle eyes :eagle: Fixed it.